crates/agglayer-node/src/rate_limiting/local/limiter/core.rs

```diff
@@ -0,0 +1,164 @@
+use super::{state::State, RawState, SlotTracker};
+
+/// Rate limiter with non-trivial state.
+///
+/// An object of [RateLimiterCore] type adds the ability to reserve rate
+/// limiting slots to the limiter state. A reserved slot takes up one space
+/// in the limiter capacity. If the maximum capacity is reached, subsequent
+/// reservation requests will be declined. A reserved slot can then be either
+/// released, freeing up one space for subsequent requests, or recorded to
+/// the rate limiter state, taking up capacity until expiration.
+#[derive(Debug)]
+pub struct RateLimiterCore<S> {
+    /// The internal state of the rate limiter.
+    state: State<S>,
+
+    /// Number of reserved slots for events being actively processed.
+    reserved: usize,
+}
+
+impl<S: RawState> RateLimiterCore<S> {
+    /// Create a new limiter state.
+    pub fn new(raw_state: S) -> Self {
+        let state = State::new(raw_state);
+        let reserved = 0;
+        Self { state, reserved }
+    }
+
+    /// Query the total occupancy, including the reserved slots.
+    pub fn query(&mut self, time: S::Instant) -> usize {
+        let num = self.state.query(time) + self.reserved;
+        assert!(num <= self.state.max_events());
```

Comment:
Should'nt it be `debug_assert` here?
What's the rational behind allowing panics here?

Reply:
> Should'nt it be `debug_assert` here? What's the rational behind allowing panics here?

I don't have a strong opinion on the use of `debug_assert` vs. plain `assert`. Here, it checks an invariant maintained by a higher-level interface, so it will not happen if the external interface is used (unless the implementation is broken).

Reply:
Changed to `log_assert`, which asserts in debug and logs an error in release.